src/rust/src/zip339_ffi.rs

```diff
@@ -0,0 +1,109 @@
+use libc::{c_char, c_uchar, c_void, size_t};
+use std::{ffi::CStr, ptr, slice};
+
+use zcash_primitives::zip339;
+
+#[repr(u32)]
+#[derive(Copy, Clone)]
+pub enum Language {
+    English = 0,
+    SimplifiedChinese = 1,
+    TraditionalChinese = 2,
+    Czech = 3,
+    French = 4,
+    Italian = 5,
+    Japanese = 6,
+    Korean = 7,
+    Portuguese = 8,
+    Spanish = 9,
+}
+
+impl From<Language> for zip339::Language {
+    fn from(language: Language) -> Self {
+        match language {
+            Language::English => zip339::Language::English,
+            Language::SimplifiedChinese => zip339::Language::SimplifiedChinese,
+            Language::TraditionalChinese => zip339::Language::TraditionalChinese,
+            Language::Czech => zip339::Language::Czech,
+            Language::French => zip339::Language::French,
+            Language::Italian => zip339::Language::Italian,
+            Language::Japanese => zip339::Language::Japanese,
+            Language::Korean => zip339::Language::Korean,
+            Language::Portuguese => zip339::Language::Portuguese,
+            Language::Spanish => zip339::Language::Spanish,
+        }
+    }
+}
+
+/// Creates a phrase with the given entropy, in the given `Language`. The phrase is represented as
+/// a pointer to a null-terminated C string that must not be written to, and must be freed by
+/// `zip339_free_phrase`.
+#[no_mangle]
+pub extern "C" fn zip339_entropy_to_phrase(
+    language: Language,
+    entropy: *const u8,
+    entropy_len: size_t,
+) -> *const c_uchar {
+    assert!(!entropy.is_null());
+    let entropy = unsafe { slice::from_raw_parts(entropy, entropy_len) }.to_vec();
+
+    if let Ok(mnemonic) = zip339::Mnemonic::from_entropy_in(language.into(), entropy) {
+        let phrase = mnemonic.phrase();
+        let phrase_len = phrase.len();
+        let phrase_cstr =
+            unsafe { libc::calloc((phrase_len + 1) as size_t, 1_usize) as *mut c_uchar };
```

Comment:
It might be better to use `CString::into_raw()` and `CString::from_raw()`, as suggested in http://jakegoulding.com/rust-ffi-omnibus/string_return/ .